File: apps/StaffApp.iOS/Sources/ErrorCodeMapper.swift
import Foundation

// MARK: - ErrorCodeMapper

enum ErrorCodeMapper {
    static func message(for error: APIError) -> String {
        switch error {
        case .missingBaseURL:
            return NSLocalizedString("error_generic", comment: "")
        case .networkError:
            return NSLocalizedString("error_network", comment: "")
        case .unauthorized:
            return NSLocalizedString("error_auth_token_expired", comment: "")
        case .decodingError:
            return NSLocalizedString("error_generic", comment: "")
        case .httpError(_, let problem):
            return messageForProblem(problem)
        }
    }

    private static func messageForProblem(_ problem: ProblemDetails?) -> String {
        guard let problem else {
            return NSLocalizedString("error_generic", comment: "")
        }
        switch problem.type {
        case "Auth.InvalidCredentials":
            return NSLocalizedString("error_auth_invalid_credentials", comment: "")
        case "Auth.AccountDeactivated":
            return NSLocalizedString("error_auth_account_deactivated", comment: "")
        case "Auth.TokenExpired":
            return NSLocalizedString("error_auth_token_expired", comment: "")
        case "Auth.InvitationNotFound":
            return NSLocalizedString("error_auth_invitation_not_found", comment: "")
        case "Auth.InvitationAlreadyUsed":
            return NSLocalizedString("error_auth_invitation_already_used", comment: "")
        case "Auth.InvitationExpired":
            return NSLocalizedString("error_auth_invitation_expired", comment: "")
        case "Auth.InvitationRoleConflict":
            return NSLocalizedString("error_auth_invitation_role_conflict", comment: "")
        case "Auth.EmailAlreadyRegistered":
            return NSLocalizedString("error_auth_email_already_registered", comment: "")
        case "Validation.Failed":
            if let errors = problem.errors {
                let messages = errors.values.flatMap { $0 }
                if !messages.isEmpty {
                    return messages.joined(separator: "\n")
                }
            }
            return NSLocalizedString("error_validation_failed", comment: "")
        case "Complaint.NotFound":
            return NSLocalizedString("error_complaint_not_found", comment: "")
        case "Complaint.MaxMediaAttachmentsExceeded":
            return NSLocalizedString("error_complaint_max_media", comment: "")
        case "Complaint.InvalidMediaType":
            return NSLocalizedString("error_complaint_invalid_media_type", comment: "")
        case "Complaint.MediaTooLarge":
            return NSLocalizedString("error_complaint_media_too_large", comment: "")
        case "Complaint.InvalidStatusTransition":
            return NSLocalizedString("error_complaint_invalid_status_transition", comment: "")
        case "Complaint.AlreadyResolved":
            return NSLocalizedString("error_complaint_already_resolved", comment: "")
        case "Staff.CannotSetBusyManually":
            return NSLocalizedString("error_staff_cannot_set_busy_manually", comment: "")
        case "Staff.NotFound":
            return NSLocalizedString("error_generic", comment: "")
        case "System.StorageUnavailable":
            return NSLocalizedString("error_system_storage_unavailable", comment: "")
        default:
            return problem.detail ?? NSLocalizedString("error_generic", comment: "")
        }
    }
}

File: apps/StaffApp.Android/app/src/main/java/com/acls/staff/ui/ErrorCodeMapper.kt
package com.acls.staff.ui

import android.content.Context
import com.acls.staff.R
import com.acls.staff.data.remote.ApiException

/**
 * Maps API error codes to user-facing string resources.
 * Falls back to the generic error message if the code is unknown.
 */
object ErrorCodeMapper {

    private val mapping = mapOf(
        "Auth.InvalidCredentials" to R.string.error_auth_invalid_credentials,
        "Auth.AccountDeactivated" to R.string.error_auth_account_deactivated,
        "Validation.Failed" to R.string.error_validation_failed,
        "Complaint.InvalidStatusTransition" to R.string.error_complaint_invalid_status_transition,
        "Complaint.MaxMediaAttachmentsExceeded" to R.string.error_complaint_max_media,
        "Complaint.InvalidMediaType" to R.string.error_complaint_invalid_media_type,
        "Complaint.MediaFileTooLarge" to R.string.error_complaint_media_too_large,
        "Complaint.NotFound" to R.string.error_complaint_not_found,
        "Staff.CannotSetBusyManually" to R.string.error_staff_cannot_set_busy,
        "System.InternalError" to R.string.error_generic,
        "System.StorageUnavailable" to R.string.error_generic,
        "System.DatabaseUnavailable" to R.string.error_generic
    )

    fun getMessage(context: Context, exception: Throwable): String {
        if (exception is ApiException) {
            val resId = mapping[exception.errorCode]
            if (resId != null) return context.getString(resId)
        }
        return context.getString(R.string.error_generic)
    }

    fun getMessage(context: Context, errorCode: String): String {
        val resId = mapping[errorCode]
        return if (resId != null) context.getString(resId) else context.getString(R.string.error_generic)
    }
}
